src/core/coap/coap.cpp

```diff
@@ -655,8 +654,9 @@ Error CoapBase::PrepareNextBlockRequest(Message::BlockType aType,
         }
 
         // Copy option
-        SuccessOrExit(error = iterator.ReadOptionValue(&optionBuf));
-        SuccessOrExit(error = aRequest.AppendOption(optionNumber, iterator.GetOption()->GetLength(), &optionBuf));
+        SuccessOrExit(error = aRequest.AppendOptionFromMessage(optionNumber, iterator.GetOption()->GetLength(),
+                                                               iterator.GetMessage(),
+                                                               iterator.GetOptionValueMessageOffset()));
```

Comment:
How about passing a `const  Option::Iterator &aIterator` to the newly added `AppendOption()` flavor directly. This way all parameters are passed toghther (no need to read and pass `optionNumber`, option length, message, offset separately).

The`AppendOption(const Option::Iterator &aIterator) will append the option read from the given `aIterator` to the message.

Reply:
Thanks for your advice. You are always welcome to change the commit. I will learn from it.
But `Option::Iterator` is invisible to `Message`, which is great for me. I'd better not invork  `Option::Iterator` in `Message`.
And all `AppendOption`-like function gives me three arguments to add, some giving three clear arguments, some using invisible default `aNumber` or `aLength `which in their name, some calculating the `aLength ` of  `aValue`, some departing the `aValue`. I'd like to keep the same style. I do agree that using `aIterator` is more simple for adding new `AppendOption()`, but I insist the style. 
I do need more reasons to change it.

Reply:
The main reason is simplicity, readability and style itself. But it does not matter. can keep it as in